src/filemanager.cpp

```diff
@@ -0,0 +1,2301 @@
+                fprintf(stderr, "cannot find linked_chunks.id = %d.\n",
+                        static_cast<int>(id));
+                return false;
+            }
+            auto next = stmt->getInt64();
+            if (next == 0) {
+                if (id != tail) {
+                    fprintf(stderr,
+                            "last item when following next is not tail.\n");
+                    return false;
+                }
+                break;
+            }
+            if (visitedIds.find(next) != visitedIds.end()) {
+                fprintf(stderr, "found cycle on linked_chunks.next = %d.\n",
+                        static_cast<int>(next));
+                return false;
+            }
+            id = next;
+        }
+        if (visitedIds.size() != static_cast<size_t>(count_linked_chunks)) {
+            fprintf(stderr,
+                    "ghost items in linked_chunks when following next.\n");
+            return false;
+        }
+    } else if (count_linked_chunks) {
+        fprintf(stderr, "linked_chunks_head_tail.head = NULL but linked_chunks "
+                        "not empty.\n");
+        return false;
+    }
+
+    if (tail) {
+        auto id = tail;
+        std::set<sqlite3_int64> visitedIds;
+        stmt = prepare("SELECT prev FROM linked_chunks WHERE id = ?");
+        if (!stmt) {
+            return false;
+        }
+        while (true) {
+            visitedIds.insert(id);
+            stmt->reset();
+            stmt->bindInt64(id);
+            if (stmt->execute() != SQLITE_ROW) {
+                fprintf(stderr, "cannot find linked_chunks.id = %d.\n",
+                        static_cast<int>(id));
+                return false;
+            }
+            auto prev = stmt->getInt64();
+            if (prev == 0) {
+                if (id != head) {
+                    fprintf(stderr,
+                            "last item when following prev is not head.\n");
+                    return false;
+                }
+                break;
+            }
+            if (visitedIds.find(prev) != visitedIds.end()) {
+                fprintf(stderr, "found cycle on linked_chunks.prev = %d.\n",
+                        static_cast<int>(prev));
+                return false;
+            }
+            id = prev;
+        }
+        if (visitedIds.size() != static_cast<size_t>(count_linked_chunks)) {
+            fprintf(stderr,
+                    "ghost items in linked_chunks when following prev.\n");
+            return false;
+        }
+    } else if (count_linked_chunks) {
+        fprintf(stderr, "linked_chunks_head_tail.tail = NULL but linked_chunks "
+                        "not empty.\n");
+        return false;
+    }
+
+    fprintf(stderr, "check ok\n");
+    return true;
+}
+
+// ---------------------------------------------------------------------------
+
+DiskChunkCache::~DiskChunkCache() {
+    if (hDB_) {
+        sqlite3_exec(hDB_, "COMMIT", nullptr, nullptr, nullptr);
```

Comment:
what if this fails?

Reply:
nothing too bad normally: the content of the transaction will be lost (that is a chunk downloaded from the network that would normally have been cached won't), but hopefully integrity will be maintained. I've added a pj_log() trace in the unlikely event that would happen